File: Task 01/linear.cpp
#include <iostream>
#include <vector>

int main()
{
    const int SIZE = 5;
    std::vector<double> x = {0, 1, 2, 3, 4};
    std::vector<double> y = {2, 3, 5, 4, 6};
    double sum_x = 0, sum_y = 0, sum_xy = 0, sum_x2 = 0;
    for (int i = 0; i < SIZE; i++)
    {
        sum_x += x[i];
        sum_y += y[i];
        sum_xy += x[i] * y[i];
        sum_x2 += x[i] * x[i];
    }
    double slope = (SIZE * sum_xy - sum_x * sum_y) / (SIZE * sum_x2 - sum_x * sum_x);
    double intercept = (sum_y - slope * sum_x) / SIZE;
    std::cout << "Linear regression predictions:" << std::endl;
    for (int i = 0; i < SIZE; i++)
    {
        std::cout << slope * x[i] + intercept << std::endl;
    }
    return 0;
}

File: Task 01/linear.rs
fn linear_regression(x: &[f64], y: &[f64]) -> (f64, f64) {
    let sum_x = x.iter().sum::<f64>();
    let sum_y = y.iter().sum::<f64>();
    let sum_xy = x.iter().zip(y.iter()).map(|(xi, yi)| xi * yi).sum::<f64>();
    let sum_x2 = x.iter().map(|xi| xi * xi).sum::<f64>();
    (
        sum_xy / sum_x2,
        (sum_y / x.len() as f64) - ((sum_xy / sum_x2) * (sum_x / x.len() as f64)),
    )
}

fn main() {
    let x = vec![0.0, 1.0, 2.0, 3.0, 4.0];
    let y = vec![2.0, 3.0, 5.0, 4.0, 6.0];
    let (slope, intercept) = linear_regression(&x, &y);
    println!("Linear regression predictions:");
    for xi in &x {
        println!("{:.2}", slope * xi + intercept);
    }
}
